WordPress/gutenberg
# useBlockPreview: Try outputting EditorStyles to ensure local style overrides are rendered

## What?


Fixes: #54956

Explores an alternative to #55241 that instead of syncing style overrides, tries to output `EditorStyles` instead. This PR fixes an issue where Duotone styles set within a Post Template block are not rendered in non-selected instances of a Query Loop within the post or site editors. It also fixes an issue where Duotone styles in Safari were not updating for these previews, either.

## Why?


Prior to this PR, for things that use `useBlockPreview`, the styles that are generated within that preview were not output anywhere. This is because `ExperimentalBlockEditorProvider` uses its own registry, so is designed to have values passed down to it, but not to propagate values outside of itself. So any calls to `setStyleOverride` would update state within the local block editor provider, and not be captured by the parent's `EditorStyles`.

The fix in this PR is to output an `EditorStyles` component within the preview, and ensure that the local-to-the-preview block editor state only contains local styles, so we avoid and duplicates between the preview and the parent block editor environment.

There is a small drawback that the `style` and `svg` tags in this PR are rendered as a direct child of the post template block's `li` element, however based on feedback on this PR and #55241, it seems that this drawback is not as bad as the difficulty of accurately syncing between `overridesStyles` states, which was explored in #55241.

This PR is currently my preferred approach to try to fix this bug.



## How?


* Try using `EditorStyles` to output styles for the `useBlockPreview`'s local styles instead of syncing as in #55241.
* In the `useBlockPreview`'s settings that get passed down to `ExperimentalBlockEditorProvider`, strip it of `styles` so that any styles applied in the parent environment are not duplicated in the child preview's editor styles. This ensures that only styles generated within that preview are output via `EditorStyles`.
* In the Duotone hook, add some logic to repaint the block that uses Duotone styles, if the current session is running Safari. While we usually like to avoid browser sniffing, this ensures that the slight performance hit is only encountered in browsers where the repaint hack is needed.

## Testing Instructions





1. Add a couple of posts with a featured image set.
2. In another post, page, or template, add a Query Loop / Post Template block containing featured images and at least 2 items per page (you can use the test markup below).
3. In the editor, prior to this PR, only the selected item within the Post Template will render Duotone styles.
4. With this PR applied, each item in the Post Template loop should render Duotone styles.
5. If you go to adjust the Duotone styles for any of the blocks in the loop, it should update in real-time (with a slight delay as styles are propagated).
6. Test adding a bunch of other images to the same post outside of the loop, and use the Duotone filter, with custom values for the background and foreground colours. There should be no clash with the styles updated from with the Post Template loop 🤞
7. Test in the site editor as well as the post editor, and a range of browsers (including Safari), and in different block configurations (e.g. Image block as a child of Row, etc)


<details>

<summary>Test markup</summary>

```html

<div class="wp-block-query">

<div class="wp-block-group has-border-color has-text-color has-link-color" style="border-color:#929df3;border-width:3px;border-radius:10px;color:#4f0808">
<div class="wp-block-columns are-vertically-aligned-center">
<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:25%"></div>



<div class="wp-block-column is-vertically-aligned-center" style="flex-basis:75%">
<h6 class="wp-block-heading">A heading</h6>


</div>
</div>
</div>

</div>

```

</details>

### Testing Instructions for Keyboard


## Screenshots or screencast 

### Before

Duotone is only applied to the selected item (as the others use a preview whose styles are not being propagated)

https://github.com/WordPress/gutenberg/assets/14988353/5930ecd6-0136-4cca-b0bf-44600aef6f64

### After

Duotone is applied to all items, and the non selected items are updated to reflect the change. Adjusting Duotone for an image outside of the Post Template loop is unaffected:

https://github.com/WordPress/gutenberg/assets/14988353/806c259b-9816-4b10-8ca5-9352aba4e9f5
